tools/c7n_azure/c7n_azure/container_host/host.py

```diff
@@ -0,0 +1,415 @@
+from c7n_azure import entry, constants
+from c7n_azure.azure_events import AzureEventSubscription, AzureEvents
+from c7n_azure.provider import Azure
+from c7n_azure.session import Session
+from c7n_azure.storage_utils import StorageUtilities as Storage
+from c7n_azure.utils import ResourceIdParser
+
+from c7n.config import Config
+from c7n.policy import PolicyCollection
+from c7n.resources import load_resources
+from c7n.utils import local_session
+
+log = logging.getLogger("c7n_azure.container-host")
+max_dequeue_count = 2
+policy_update_seconds = 60
+queue_poll_seconds = 15
+jitter_seconds = 10
+queue_timeout_seconds = 5 * 60
+queue_message_count = 5
+
+
+class Host:
+
+    def __init__(self):
+        logging.basicConfig(level=logging.INFO, format='%(message)s')
+        log.info("Running Azure Cloud Custodian Self-Host")
+
+        if not Host.has_required_params():
+            return
+
+        load_resources()
+        self.session = local_session(Session)
+
+        # Load configuration
+        self.options = Host.build_options()
+        self.policy_storage_uri = os.getenv(constants.ENV_CONTAINER_POLICY_STORAGE)
+        self.event_queue_name = os.getenv(constants.ENV_CONTAINER_EVENT_QUEUE_NAME)
+        self.event_queue_id = os.getenv(constants.ENV_CONTAINER_EVENT_QUEUE_ID)
+
+        # Prepare storage bits
+        self.policy_blob_client = None
+        self.blob_cache = {}
+        self.queue_storage_account = self.prepare_queue_storage(
+            self.event_queue_id,
+            self.event_queue_name)
+
+        self.queue_service = None
+
+        # Track required event subscription updates
+        self.require_event_update = False
+
+        # Policy cache and dictionary
+        self.policy_cache = tempfile.mkdtemp()
+        self.policies = {}
+
+        # Configure scheduler
+        self.scheduler = BlockingScheduler()
+        logging.getLogger('apscheduler.executors.default').setLevel(logging.ERROR)
+
+        # Schedule recurring policy updates
+        self.scheduler.add_job(self.update_policies,
+                               'interval',
+                               seconds=policy_update_seconds,
+                               id="update_policies",
+                               next_run_time=datetime.now())
+
+        # Schedule recurring queue polling
+        self.scheduler.add_job(self.poll_queue,
+                               'interval',
+                               seconds=queue_poll_seconds,
+                               id="poll_queue")
+
+        self.scheduler.start()
+
+    def update_policies(self):
+        """
+        Enumerate all policies from storage.
+        Use the MD5 hashes in the enumerated policies
+        and a local dictionary to decide if we should
+        bother downloading/updating each blob.
+        We maintain an on-disk policy cache for future
+        features.
+        """
+        if not self.policy_blob_client:
```

Comment:
this feels like something better to decompose k8s pod style, ie structuring as a standard daemon would mean this comes in and reloads things from disk on unix signal. ie unix style also means do one thing and do it well. ie.  pulling policies from a git repo, copying them in a blob store and retrieving to a local fs really a separate thing entirely from listening to events and executing policies.

Reply:
Given what this is doing is pretty light weight I am trying to avoid too much complexity in the deployment (e.g. multiple pods working together).  Certainly could have a scheduler that inspects management groups and deploys worker pods per sub, a service to sync policies, etc - but my hope was to have it be dead simple so it can run anywhere with minimal configuration.

(k8s and ACI support bits coming soon) 

Regardless I take your point, at a minimum the code could be easily refactored to separate these concepts and I think it should be.